src/H5Dchunk.c

```diff
@@ -3105,6 +3105,16 @@ H5D__chunk_read(H5D_io_info_t *io_info, H5D_dset_io_info_t *dset_info)
         ctg_io_info.dsets_info   = &ctg_dset_info;
         ctg_io_info.count        = 1;
 
+        /* Ensure dataset I/O sieve buffer is properly setup for when the contiguous
+         * dataset I/O routines are used for chunks. Force reset of sieve buffer here
+         * on each I/O until more advanced use cases like H5Dset_extent between writes
+         * / writes and reads can be supported.
+         */
+        dset_info->dset->shared->cache.contig.sieve_buf_size = H5F_SIEVE_BUF_SIZE(dset_info->dset->oloc.file);
```

Comment:
This line (and the same one in `H5D__chunk_write()`) is the main fix. The library's logic seems to have been setup to account for chunked datasets making use of data sieving, but the sieve buffer size was always 0 so it was unused. In specific cases this could result in I/O being performed element by element.